Bluetooth / Radio communication between bots!

Woonters opened the pull request:
This should close #52 

There's still some stuff I think might be worth doing:

[ ] - Review current message implementation, is the message size, style etc. worth it?
[ ] - Calculate the cooldown values and adjust them a for balancing
[ ] - Should the write buffer have better write functionality? I would consider using an offset from MEM_BLUETOOTH and writing each 4 bytes instead of just sending 1 byte at a time, this would probably mean you should be able to read from it as well, effectively treat it like a memory location 
[ ] - Check over all the names for interfacing functions in kartoffel::bluetooth
[ ] - Add wider / better tests 
[ ] - Tutorial level 

## Other notes

I added two test bots Ianthe and Coronabeth, happy to delete them, but thought if I was going to do a tutorial level they might be useful currently they just work as a tx rx pair doing nothing else 

I'd certainly be interested in other messaging concepts as well, creating a send / receive stream between two bots for instance could be a fun task

I think when writing this there was a small part of my brain always thinking about electronic warfare, it'd be cool if you could do some sort of that in a way, but also probably really messy, I made the bots send their own ID as part of the message, but if you wanted to encourage people to write their own *clever* networking code then maybe the bot id section could be dropped (and i'd probably increase the Message length to accommodate

Patryk27 committed:
backend: Get rid of world events

Patryk27 committed:
tutorial: Adjust to recent changes

Patryk27 committed:
world: Show just couple first characters of hash

Patryk27 committed:
acyclic-maze: Refactoring

Patryk27 committed:
flight-syndome: Start working on challenge

Patryk27 committed:
session: s/single_bot_mode/hero_mode

Patryk27 committed:
session: Improve camerawork

Patryk27 committed:
ui: Implement fade-in / fade-out effects

Patryk27 committed:
session: Use lower-cased statuses

Patryk27 committed:
radar: Support bot ids

Patryk27 committed:
session: Refactoring

Patryk27 committed:
world: Fix a crash when a queued bot is restarted

Patryk27 committed:
flight-syndrome: Continue working on challenge

Patryk27 committed:
ui: Refactoring

Patryk27 committed:
game: Continue working on prefabs

Patryk27 committed:
game: Continue working on prefabs

Patryk27 committed:
tests: Refresh snapshots

Patryk27 committed:
session: Fix caret

Patryk27 committed:
ui: Make more haxxxy

Patryk27 committed:
game: Continue working on prefabs

Patryk27 committed:
ui: Rephrase couple of things

Patryk27 committed:
session: Improve the `spawn-bot` command

Patryk27 committed:
kartoffel: Use NonZeroU64 to represent bot ids

Patryk27 committed:
Simplify managing prefabs

Patryk27 committed:
readme: Update

Patryk27 committed:
readme: Update

Patryk27 committed:
hostage-situation: Continue working on challenge

Patryk27 committed:
bots: Add comments

Patryk27 committed:
just: Add `perf` command

Patryk27 committed:
cpu: Simplify

Patryk27 committed:
diamond-heist: Continue working on challenge

Patryk27 committed:
world: Drop dependency on `lru`

Patryk27 committed:
kartoffel: s/tile_at/at

Patryk27 committed:
cpu: Implement `Debug` by hand

Patryk27 committed:
server/ssh: Specify `auth_banner`

Patryk27 committed:
server/ssh: Force compression

Patryk27 committed:
world: Adjust terminology around snapshot's map

Patryk27 committed:
world: s/destroy_bot/delete_bot

Patryk27 committed:
world/inventory: Drop `#[serde(transparent)]`

Patryk27 committed:
diamond-heist: Wait a moment before driving out

Patryk27 committed:
diamond-heist: Start implementing objects

Patryk27 committed:
tests/tutorial: Make more stable

Patryk27 committed:
world: Simplify bot management

Patryk27 committed:
session: Start working on the inspect-bot command

Patryk27 committed:
session/game: Improve camera

Patryk27 committed:
world: Adjust `Dir::fmt()` to print lowercased values

Patryk27 committed:
world: Change snapshot broadcasting duration (50ms -> 33ms)

Patryk27 committed:
world: Add rudimentary tests for events

Patryk27 committed:
world: Adjust kill messages

Patryk27 committed:
world: Fix `AliveBots::remove()`

Patryk27 committed:
session: Refactoring

Patryk27 committed:
session/menu: Change button labels

Patryk27 committed:
session/game: Refactoring, bring back previous go-back modal

Patryk27 committed:
challenges: Provide dedicated key codes

Patryk27 committed:
acyclic-maze: Improve dialogs

Patryk27 committed:
tests: Replace `â` with `#`

Patryk27 committed:
tests/tutorial: Make more reproducible

Patryk27 committed:
diamond-heist: Complete challenge

Patryk27 committed:
world/inventory: Use `VecDeque`

Patryk27 committed:
kartoffel: Add docs for `arm_pick()` and `arm_drop()`

Patryk27 committed:
world/migrations: Fix v13's test

Patryk27 committed:
just: s/start-demo/demo

Patryk27 committed:
nixos: Disable `kartoffels.backend.debug`

Patryk27 committed:
ui: Simplify the "terminal is too small" error

Patryk27 committed:
frontend: Improve error messages

Patryk27 committed:
Compress WebSocket messages

Patryk27 committed:
session: Refactoring

Patryk27 committed:
world: Adjust logs

Patryk27 committed:
chore: Bump toolchain

Patryk27 committed:
tutorial: Improve completion message

Patryk27 committed:
tests/tutorial: Don't assert `completed.txt`

This assertion is not reproducible. It looks like a bug in avt, but I
don't have time to investigate it more at the moment.

Patryk27 committed:
cyclic-maze: Start working on the challenge

Patryk27 committed:
session: Improve map-revealing algorithm

Patryk27 committed:
challenges: Center call to actions

Patryk27 committed:
kartoffel: Add `timer_wait()` function

Patryk27 committed:
world: Add bot-logs for edge cases

Patryk27 committed:
personal-roomba: Continue working on the challenge

Patryk27 committed:
kartoffel: Fix `timer_wait()`'s docs

Patryk27 committed:
world/migrations/v13: Migrate compass

Patryk27 committed:
session: s/get_snapshot_version/get_world_version

Patryk27 committed:
session: s/move_at/look_at

Patryk27 committed:
tests: Avoid replacing `â` with `#`

Patryk27 committed:
tests: s/not_see/dont_see

Patryk27 committed:
Use em dashes where applicable

Patryk27 committed:
game: Don't follow bots when testing

Patryk27 committed:
fmt(game): Use let-chains

Patryk27 committed:
world/compass: Fix `mmio_load()`

Patryk27 committed:
world/inventory: Fix tests

Patryk27 committed:
world: Support events

Patryk27 committed:
diamond-heist: Adjust `GUARD_KILLED_MSG`

Patryk27 committed:
Drop `robot combat arena` from the slogan

Patryk27 committed:
frontend: Add description meta

Patryk27 committed:
kartoffel: Get rid of `inv_len()`

Patryk27 committed:
chl/chl-personal-roomba: Apply lints

Patryk27 committed:
readme: Refresh images

Patryk27 committed:
readme: Center text

Patryk27 committed:
diamond-heist: s/arm_take/arm_pick

Patryk27 committed:
session: Remove debug mode

Patryk27 committed:
chore: Release v0.6.0

Cryptjar committed:
kartoffel: Add fmt::Write impl and output the panic message on the serial.

Cryptjar committed:
Put the panic printing behind a default feature.

Patryk27 committed:
Merge pull request #40 from Cryptjar/impl-panic

Write the panic message to the serial output

Patryk27 committed:
session/index: Adjust text in header

Closes #41

Patryk27 committed:
frontend: Avoid async `event.data.arrayBuffer()`

Closes #42

Patryk27 committed:
nix(backend): Set `__darwinAllowLocalNetworking`

Patryk27 committed:
toolbox: Implement `create-world` command

Patryk27 committed:
store: Adjust `Store::dir()` panic msg

Patryk27 committed:
cpu: Implement 32-bit atomics

Related to #43; 64-bit atomics are next.

Patryk27 committed:
cpu: Move benchmark to `benches` directory

Patryk27 committed:
tests(backend): s/home/index

Patryk27 committed:
cpu: Implement amo{min/max}u.w

Patryk27 committed:
tests(cpu): Improve assertion message

Patryk27 committed:
cpu: Remove extra `& 0xff`

Patryk27 committed:
benches(cpu): Simplify benchmark

Patryk27 committed:
fix(cpu): Don't read s_imm for sc.w

Patryk27 committed:
tests(cpu): Refactor and improve the xx-ints test

Patryk27 committed:
tests(cpu): Use 64-bit constants for 32-bit tests

Doing so makes the tests more thorough, just in case the emulator
accidentally uses 64-bit logic somewhere.

Patryk27 committed:
cpu: Implement 64-bit atomics

Closes #43

Patryk27 committed:
backend: Enable TCP_NODELAY for SSH connections

Patryk27 committed:
backend: Patch-out cryptovec's mlock

NixOS containers (so systemd-nspawn) disables mlock by default,
panicking all SSH connections.

Patryk27 committed:
refactor(world): Use bevy_ecs

Patryk27 committed:
world: Use ChaCha instead of SmallRng

Patryk27 committed:
chore: Bump dependencies

Patryk27 committed:
Merge pull request #44 from Patryk27/bevyfication

refactor(world): Use bevy_ecs

Patryk27 committed:
play: Mention that pause is local-only

Patryk27 committed:
backend: Fix `cargo build-cpu-tests`

Patryk27 committed:
backend: Fix `cargo test -p kartoffels-world`

Patryk27 committed:
backend: Start working on runs & statistics

Patryk27 committed:
backend: Continue working on runs & statistics

Patryk27 committed:
backend: Reexport termwiz under kartoffels_ui

Patryk27 committed:
backend: s/TermEndpoint/TermFrontend

Patryk27 committed:
ui: Simplify area management

Patryk27 committed:
backend: Bless tests

Patryk27 committed:
play: Close joining modal when given empty id

Patryk27 committed:
backend: Continue working on statistics

Patryk27 committed:
backend: Use let-chains more

Patryk27 committed:
session: Extend bots modal

Patryk27 committed:
world: Change save duration to 15 mins

Patryk27 committed:
session: Show an extra note if bot has been alive for (too) long

Patryk27 committed:
backend: s/runs/lives

Patryk27 committed:
session: Simplify bots modal

Patryk27 committed:
refact: s/backend/app, s/frontend/web

Patryk27 committed:
nix: Fix russh-cryptovec-mlock.patch

Patryk27 committed:
app: s/kartoffels-bots/kartoffels-prefabs

Patryk27 committed:
app: s/kartoffels-session/kartoffels-frontend

Patryk27 committed:
app: Simplify default.nix

Patryk27 committed:
frontend/inspect-bot: Clarify message

Patryk27 committed:
kartoffel: Mention that timer overflows after 18h

Patryk27 committed:
fix(world/lives): Iterate past lives in reverse

Patryk27 committed:
fix(world): Distinguish wall-time from world-time

Patryk27 committed:
frontend: Show `pos: x,y` instead of `pos: [x, y]`

Patryk27 committed:
just: s/test-bless/test-and-bless

Patryk27 committed:
stats: Collect stats for ages

Patryk27 committed:
frontend: Use bg color to distinguish panels

Patryk27 committed:
world: Optimize

Patryk27 committed:
inspect-bot: Improve modal

Patryk27 committed:
inspect-bot: Add charts

Patryk27 committed:
Merge pull request #45 from Patryk27/stats

backend: Implement statistics

Patryk27 committed:
kartoffel: Fix typo

Patryk27 committed:
world/events: s/while/if

Patryk27 committed:
kartoffel: Improve top-level docs

Patryk27 committed:
Improve serial port, provide print!() macros

Closes #50

Patryk27 committed:
kartoffel/panic: Refactoring

Patryk27 committed:
kartoffel/serial: Improve docs

Patryk27 committed:
kartoffel/radar: Refactor API (#49)

Patryk27 committed:
kartoffel: Make RadarScan non-user-constructible

Patryk27 committed:
inspect-bot: s/your/this

Patryk27 committed:
world: Add more snapshot tests

Patryk27 committed:
tests: Bless snapshots

Patryk27 committed:
cpu: Migrate from rv64 to rv32

Patryk27 committed:
world: Remove unnecessary migrations

Patryk27 committed:
cpu: Migrate to riscv32

Closes #53

Patryk27 committed:
world: Augment `fmt::Debug` with `tiles.hash`

Patryk27 committed:
admin: Start implementing

Patryk27 committed:
game: Let spawn-bot default to prefab.roberto

Patryk27 committed:
game: Simplify bot source management

Patryk27 committed:
ui: Refactoring

Patryk27 committed:
admin/create-world: Start implementing command

Patryk27 committed:
frontend/error: Use `err.to_fmt_string()`

Patryk27 committed:
world: Remove unused file

Patryk27 committed:
admin/create-world: Implement command

Patryk27 committed:
console: Minor refactoring

Patryk27 committed:
store: Refactor world management

Patryk27 committed:
console: Implement more commands

Patryk27 committed:
store: Implement more tests

Patryk27 committed:
console/exit: Implement command

Patryk27 committed:
world/policy: Improve FromStr impl

Patryk27 committed:
play: Enable debug-mode for admin sessions

Patryk27 committed:
world: Support caves instead of dungeons

Patryk27 committed:
justfile: Extract `demo` into `create-demo` and `update-demo`

Patryk27 committed:
nix: Adjust artifact names

Patryk27 committed:
world: Add more tests

Patryk27 committed:
nixos: Support `--secret`

Patryk27 committed:
button: Swap places of `label` and `key` args

Patryk27 committed:
button: Refactoring

Patryk27 committed:
play: Refactor view

Patryk27 committed:
frontend: Refactor fade-ins and fade-outs

Patryk27 committed:
tutorial: Assert the last message as well

Patryk27 committed:
frontend: Preinitialize background widget

Patryk27 committed:
fade: Modulate both foreground and background

Patryk27 committed:
Merge pull request #54 from Patryk27/caves

world: Support caves instead of dungeons

Patryk27 committed:
motor: Support backward movement

Patryk27 committed:
docs: Fix typo

Patryk27 committed:
sandbox: Adjust predefined cave sizes

Patryk27 committed:
admin/rename-world: Implement command

Patryk27 committed:
admin/create-world: Make `name` position argument

Patryk27 committed:
store: Avoid `#[instrument]` for `Worlds::create()`

Patryk27 committed:
world: Improve lifecycle logs

Patryk27 committed:
chore: Apply Clippy suggestions

Patryk27 committed:
refact: Use `.store()` instead of `.swap()`

Patryk27 committed:
store: Make `Worlds::list()` stable

Patryk27 committed:
console: Keep stdin and stdout next to each other

Patryk27 committed:
tests: Create tests for console

Patryk27 committed:
app: Clean-up terminology (s/robot/bot)

Patryk27 committed:
app: s/data/store

Patryk27 committed:
readme: Provide building instructions

Patryk27 committed:
readme: Sync

Patryk27 committed:
store: Avoid `#[instrument]` for `Store::load()`

Patryk27 committed:
inspect-bot: Fix the `curr-life` counter

Patryk27 committed:
chore: Bump toolchain and dependencies

Patryk27 committed:
chore: Release v0.7.0

Patryk27 committed:
readme: Update images

Patryk27 committed:
upload-bot: Fix `./build --upload`

Woonters committed:
Update is_pos_legal to have an air-gap of 1 for potential spawns

Woonters committed:
fix and optimise is_pos_legal changes

switched around check orderings to reduce likelyhood of checking all the squares if the spawn one is invalid
fixed the x and y loops to include 1 actually creating the circle, i'm always forgetting inclusivity / exclusivity on for loops!
removed duplicate bots check

Patryk27 committed:
kartoffel: Skip build.rs on non-riscv32 archs

Patryk27 committed:
kartoffel: Release on crates.io

Woonters committed:
update is_pos_legal to only check surrounding area in certain instances

Patryk27 committed:
Merge pull request #59 from Woonters/main

Update is_pos_legal to have an air-gap of 1 for potential spawns

Patryk27 committed:
world: Bless tests

Woonters committed:
Outline of bluetooth functionality in kartoffels-world

Woonters committed:
first version of bluetooth module in kartoffels-world

Woonters committed:
create bluetooth api for kartoffel crate

Woonters committed:
Fix to is_pos_legal to stop bots spawning on the same tile if they aren't checking neighborhood

this is a necessary change since the hashmap AliveBots.pos_to_id means that two bots on the same position could break bot lookup code

Patryk27 committed:
Merge pull request #60 from Woonters/main

Fix to is_pos_legal to stop bots spawning on the same tile

Woonters committed:
rewite the message buffer and core read functionality

Woonters committed:
rewrite bluetooth interface fix some bugs and create two test bots

Woonters committed:
document and clean up bluetooth files

Woonters committed:
Tests for messages and message buffer

Patryk27 commented:
Thanks for the pull request - I like the direction and, having thought about it for a couple of days, I think we could take it a couple of steps further!

Note that everything below is a proposal, a bag of thoughts - feel free to comment, come up with counter-ideas, and so on.

---

Overall, I tend for each peripheral to be somewhat different from each other not only in the tangible functionality, but also the intrinsic mechanic -- serial drops older bytes, compass auto-refreshes in the background etc.

For Bluetooth (aka radio, and we'll get to the naming soon) the mechanic we could apply to make it intrinsically different from radar is to make it possible to send messages _without_ specifying the radius. Rather, the message travels _outwards_ from the bot and with every tick it travels, it wanes, increasing the chance for observing bit flip(s) within it.

(in a similar spirit -- when two messages collide mid-air, both can get mangled, which touches on the electronic warfare subject you mentioned.)

Practically speaking, instead of having `send_bluetooth(radius)`, we'd have `radio_send(msg)` that emits given message outwards, propagating it with some constant speed, say 1 tile every 256 world-ticks:

```
  .
```

->

```
  .
 . .
  .
```

->

```
  .
 . .
.   .
 . .
  .
```

When a message "hits" a bot, that's when it's pushed onto the radio's "message stack" (if there's space for it), with some of the bits mangled (maybe 1 bit for each 1 traversed tile? though it would certainly make much more sense to apply inverse square law here so that messages received in close proximity have almost 100% chance of being correct).

## Naming

I'm feeling _radio_ more than _Bluetooth_ - that's because Bluetooth kinda implies a high-level thing going on (with device discovery etc. provided by the protocol), while what we'd like to offer is more of a "send a bag of bytes and receive a bag of bytes" functionality.

In particular, I'd like to incentivize players to implement error recovery algorithms (Hamming code and whatnot) and higher-level protocols on top of the basic tools provided for the game.

## Interface

As with everything in here, the interface here is also a proposal - a sketched version of my thoughts is:

```rust
// N.B. we could assume that radio is always turned on, but at some point I'd
//      like to revisit the concept of implementing a battery - turning the 
//      radio on/off could be a battery-saving strategy then
pub fn radio_on() {
    // N.B. we're reserving 0x01 as a general "set radio power" command - in the
    //      future we could have a "set_radio_power()" function that would cause
    //      the radio to eat more battery in exchange for allowing it to send
    //      messages for longer distances (and vice versa - reduce the chance of
    //      observing a bit flip in a retrieved message)
    //
    //      this would be then implemented as cmd(0x01, power, 0x00, 0x00)
    wri(MEM_RADIO, 0, cmd(0x01, 0x01, 0x00, 0x00));
}

pub fn radio_off() {
    wri(MEM_RADIO, 0, cmd(0x01, 0x00, 0x00, 0x00));
}

pub fn radio_status() -> u32 {
    rdi(MEM_RADIO, 0)
}

pub fn is_radio_on() -> bool {
    radio_status() & 1 > 0
}

// aka "can we send the next packet"
pub fn can_radio_send() -> bool {
    radio_status() & 2 > 0
}

// aka "is there any packet waiting for us in the queue"
pub fn can_radio_recv() -> bool {
    radio_status() & 4 > 0
}

pub fn radio_send(msg: &[u8; 32]) {
    for (idx, payload) in msg.array_windows().enumerate() {
        wri(MEM_RADIO, idx + 1, u32::from_le_bytes(payload));
    }

    wri(MEM_RADIO, 0, cmd(0x02, 0x00, 0x00, 0x00));
}

// if there's no packet, just fills msg with zeros
pub fn radio_recv(msg: &mut [u8; 32]) {
    for (idx, payload) in msg.array_windows_mut().enumerate() {
        *msg = rdi(MEM_RADIO, idx + 1).to_le_bytes();
    }
}

// Shiny thing! Allows to specify a predicate for radio_recv().
//
// E.g. calling `radio_filter(&[0xca, 0xfe, 0xba, 0xbe])` will cause the radio
// to transparently ignore all incoming messages which have a different prefix;
// doesn't affect `radio_send()`.
//
// This pairs up with `radio_send()` *not* allowing to send a message to a
// specific robot id -- rather, players are encouraged to come up with a
// "namespacing" scheme of their own.
//
// For instance - if someone wants to have a direct bot-to-bot connection, they
// can prefix their messages with the bot id (or rather a hash of it / half of
// it, since we purposefully keep the prefix shorter than a full bot id).
//
// Alternatively - if someone wants to form a bot-cluster and communicate within
// it instead of directly to a specific bot, they can just randomize a 32-bit
// number and use it to prefix the messages.
//
// Even more combinations exist, e.g. the prefix could encode a message type so
// that you can have different bots deployed on the field that respond respond
// only to their own specific message type (which is sort of a variation on the
// clustering mechanism above, now that I think about it).
pub fn radio_filter(prefix: &[u8; 4]) {
    wri(MEM_RADIO, 0, cmd(0x03, 0x00, prefix[0], prefix[1]));
    wri(MEM_RADIO, 0, cmd(0x03, 0x01, prefix[2], prefix[3]));
}
```

## Tutorial

I think the tutorial is plenty long as it is - instead of introducing more chapters (which could be frustrating for people who want to just grasp the basics), we could introduce a new challenge.

I was thinking a distributed key-value database? :eyes:

You'd have to implement a couple of bots that together form a cluster that's able to respond to basic queries such as "put foo=bar" or "get foo" (those would also get transmitted to your robot over the radio from a built-in challenge-robot).

You'd be forced to implement more than one robot simply because the challenge-robot would generate over 128 kB of random queries - so you would _have_ to keep multiple robots working together, so that each robot contains its own "set" (shard?) of the database, otherwise you wouldn't be able to respond back to all of the queries asked by the challenge-bot.

In the extreme version we could force the robots to fight with some enemies - this way you'd be forced to implement some kind of storage redundancy; but that's maybe for another day.

## Open questions

### radio_filter() vs bitrot

If messages are applicable to this "distance bitrot", then it would make sense for each bit of the message to be _equally likely_ to get flipped - and this can make it frustrating to use `radio_filter()`, because when configured for filter `0xcafebabe`, it would reject a message that begins with `0xcaffbabe`, even though that message could well be meant for the robot.

(i.e. the specific problem is that `radio_filter()` isn't error-recovery-aware - and it cannot be, since it'd be implemented within the game itself)

A non-realistic (I think?) but good-enough solution could be to penalize further bits more than the initial bits, something like:

```rust
fn mangle_message(msg: &mut [Bit; 256], prob: f32) {
    for (idx, bit) in msg.iter_mut().enumerate() {
        let bit_prob = (*idx as f32 / 128.0).clamp(0.0, 1.0) * prob;
        
        if rng.gen(bit_prob) {
            *bit = bit.flip();
        }
    }
}
```

Intuitively, when calling `mangle_message(..., 0.1)` we'd expect for 10% of the bits to be mangled, just skewed towards the end of the message instead of uniformly throughout it; this wouldn't make `radio_filter()` bulletproof, but it'd make it more fun and practical to use, I think.

### Message size vs cooldown vs challenge 

If we're able to send just 32b of payload, then the whole "challenge requires you to handle hundredths of kilobytes of queries" part might take solid minutes to complete, mostly waiting for the queries to be sent/received (= no fun).

On the other hand, forcing messages to be e.g. 128b can also be wasteful (those bytes need to be copied into MMIO, after all).

Maybe a good compromise would be to allow to send messages of arbitrary length? (but more than 4b, to match the prefix, and less than 128b, not to overflow the receiving part)

This would be sort of "pay as you go" mechanism: you want to send a short message? cool; you want to send a longer message? cool, but you have to wait a bit longer.

## Next steps

Alright, this turned out to be quite a message! Feel free to treat is as an RFC - nothing here is set in stone, it's just a couple of ideas.

In terms of the implementation, nothing here shouts _terribly difficult_ to me - if you'd want to give it a stab, I'd love to support and review! If you don't have much time or ain't feeling like implementing more at the moment, that's also fine, just lemme know.

(but as I said, I think it's a nice, moderately challenging project that should be also quite rewarding; I've got a couple of blog posts to write myself, so I can't complain for lack of work at the moment either :smile:)

### Implementation hints

Assuming that the basic idea here stands, we'd have to store messages into the world itself:

https://github.com/Patryk27/kartoffels/blob/9e7a6f90bfeb7a5ff974fdb00d1c5af13b903dcf/app/crates/kartoffels-world/src/lib.rs#L167

Something akin to:

```rust
struct Messages {
    message: Vec<Message>,
}

struct Message {
    emitted_at: IVec2,
    radius: i32,
}
```

Then we'd have to create a new system be responsible for propagating those messages, checking when they collide with bots, and putting them into bots' recv queues (performing the `radio_filter()` check, bitrot mechanism etc.):

https://github.com/Patryk27/kartoffels/blob/9e7a6f90bfeb7a5ff974fdb00d1c5af13b903dcf/app/crates/kartoffels-world/src/lib.rs#L265

Something like:

```rust
fn receive_messages(messages: ResMut<Messages>, bots: ResMut<AliveBots>) {
    let mut msgs_to_delete = Vec::new();
    
    for (msg_idx, msg) in messages.iter_mut().enumerate() {
        for bot in bots {
            // Having `O(n * m)` isn't ideal, especially since this comparison
            // will necessarily return _false_ most of the time, but it's a good
            // starting point
            if bot.pos.distance(msg.emitted_at) == message.radius {
                bot.msgs.receive(msg);
            }
        } 
        
        msg.radius += 1; // possibly every 128 world ticks or so, depending on what's more fun in practice

        if msg.radius >= 32 { // possibly even further apart, depending on how it turns out in practice
            msgs_to_delete.push(msg_idx);
        }
    }
    
    /* ... */
}
```

Eventually messages should also be persisted into the savefile:

https://github.com/Patryk27/kartoffels/blob/9e7a6f90bfeb7a5ff974fdb00d1c5af13b903dcf/app/crates/kartoffels-world/src/storage/systems/save.rs#L71

... but for development you can start off with `Messages::default()`.

Woonters commented:
Hi really like all of these ideas, and I think you've hit the nail on the head with practically all of them
I'm happy to go about implementing all of this, only really with one or two questions.
## 1.
If we are doing the different message sizes from 4 to 128 Bytes then should the recieve buffer still be 5 messages long? 
I think the filter mainly deals with the issue of message spam, but still it feels a shame that a radio module would have it's buffer filled all the same from 20B and 512B
## 2.  
With the comment on message mangling:
> "(in a similar spirit -- when two messages collide mid-air, both can get mangled, which touches on the electronic warfare subject you mentioned.)"

Would it be worth considering the *strength* of the sent message as to how much it mangles / gets mangled

Woonters committed:
Outline for Messages in world

Woonters committed:
Rename bluetooth elements to radio

Woonters committed:
Read functionality for radio module

Woonters committed:
Update read functionality for status and add basic write functionality again

Woonters committed:
Add basic tests for the message buffer

Woonters committed:
Add the basic send functionality for messages and most of the code for messages expanding etc.

Woonters committed:
clean some function returns and document radio

Woonters committed:
move the message buffer to its own file and rework to make read work with available addressable space

Woonters commented:
Hello! 
Thought I'd check in with what's been happening so far.
It's come on leaps and bounds, but there are some implementation choices that I'm spending a bit of time thinking about

## The message buffer 
Originally I tried to make the reading of the message buffer quite transparent. Basically the first 132 Bytes are for the data about the actual radio module, and the message it's getting ready to send. 
| radio data | message we are going to send |
| --- | --- |
| 0 | 4..132 | 

Then the next chunk of addresses were for the message buffer.
I tried using a circular buffer, similar to my old implementation, but now with variable message lengths, we need to store pointers to each of the messages in the buffer and how long each one is.

Originally I was aiming for 512B of messages, meaning you could hold 4 max size messages
and 128 minimum size messages!

<table>
<tr>
<th colspan="2"> message address </th>
<th> length </th>
<th> unused </th>
</tr>
<tr>
<td>Byte</td><td>Byte</td><td>Byte</td><td>Byte</td>
</tr>
</table>

with 512 addressable bytes in the buffer we would need to store a message pointer as a u16, and length can be a u8. 
This means that realistically each message pointer is 4 bytes itself, meaning the pointer list can be 512B as well! (as seen above)
Here we have a problem, with just the message buffer itself we would need 1028B of addressable space (more than any other module uses) and that doesn't include the rest of the radio data.

| message buffer info | message pointers | message buffer
| --- | --- | --- |
| 0 | 4..516 | 516..1028 | 

or in total

| radio data | message we are going to send | message buffer info | message pointers | message buffer |
| --- | --- | --- | --- | --- | 
| 0 | 4..132 | 132| 136..664 | 664..1176 |

An alternative is just presenting the top message on the buffer, this means that we only need to present 132B of data for the message buffer, but this reduces how much the user can interact with the message buffer.
If the user wants to do some logic thinking about multiple messages then they would need to copy messages into their own structures in memory which feels a little less efficient and clean.

The status of the radio module could then also send the length of the current message at the top of the buffer

Or we can limit the number of messages pointers we can store, this would mean that there could possibly be unused space in the message buffer if lots of 4B messages are sent

I've included solutions for both ways so it should be easy to play around and choose one, (though currently the read functionality for the presenting only the top message) has some really horrible slice manipulation / work that I want to go back to and clean asap

## The message emitting
I've also been actually writing the code for the messages in world space, and already come into some fun weird interactions!

The mangling idea is interesting, but if we model a message as 
https://github.com/Patryk27/kartoffels/blob/04ce002b20562b3f401ce79dce21e2809457c53c/app/crates/kartoffels-world/src/messages.rs#L103-L112
then two messages `a` and `b` that have an intersecting point `c` will mangle
```
  a b
 a c b
  a b
```
then even bots touching `a` on the left edge will receive a message mangled by b 
this wouldn't make much sense... So the content of a message needs to be linked to each position
Also when the message expands next does the mangled version expand out or is it returned to the orignial state i.e. (ignoring now the `b` message
```
  a
 a c
a   c
 a c
  a
```
or just 
```
  a
 a a
a   a
 a a
  a
```
 This follows nicely onto the expanding message shape / code
 I've tried implementing some way of working out the next positions in a message as it expands and come up with this little algorithm
 https://github.com/Patryk27/kartoffels/blob/04ce002b20562b3f401ce79dce21e2809457c53c/app/crates/kartoffels-world/src/messages.rs#L132-L155
 Frankly I'm not happy with it, it feels messy and if the mangling being location dependent would also break the functionality I think
 I'm going to have a bit of a look around, see if I can work out a nicer implementation
 
 Finally in this part, (mostly as a note for myself) the check for if a message should be sent to a bot only happens when it expands, this could in theory mean a bot could miss a message by driving onto a cell whilst a message is there. 
 
 ## Conclusions / TLDR:
 - I'd appreciate thoughts on how the message buffer should be presented to the bot, should it just be effectively a *raw* representation (if so what do we limit to bring the size down) or should it just present the front value of the buffer?
- How  would we like to deal with message mangling? Do messages that hit each other continue to be mangled as they expand out, or is it just where they intersect?
- Any big problems with my implementations so far? I know my code is pretty messy, I'm looking to clean up some signatures, get rid of some nasty lines that read poorly and probably reduce the comments down to something more reasonable

Patryk27 commented:
Hi, amazing! I'll take a look at it tomorrow 🦀

Patryk27 commented:
> I'd appreciate thoughts on how the message buffer should be presented to the bot, should it just be effectively a raw representation [...] or should it just present the front value of the buffer?

How about we level the playing field?

Instead of having explicit `message_ptrs: Vec<MessagePointer>` etc., we could go with (in sorta pseudocode):

```rust
struct BotRadio {
    // Radio's memory.
    //
    // memory[0] is ignored - it would contain the status byte, but it's better
    // to materialize it on the fly (see BotRadio::mmio_read())
    //
    // memory[1] contains the length of the first message, memory[2] contains
    // the first four bytes of that message's payload etc.; if memory[1] == 0,
    // there's no message waiting to be received
    //
    // memory[2 + memory[1]] contains the length of the second message (possibly
    // zero if there's no second message waiting to be read) and similarly the
    // bytes that follow contain that message's payload etc.
    //
    // Messages essentially form a linked list, but instead of having `next`
    // as a pointer (which would take an extra `u16`), all messages are located
    // one after another:
    //
    // mem[0] | mem[1]          | mem[2..]            | mem[..]          | ...
    // -------|-----------------|---------------------|------------------|-----
    // xxxxxx | first msg's len | first msg's payload | second msg's len | ...
    memory: Vec<u32>,
}

impl Default for BotRadio {
    fn default() -> Self {
        Self {
            memory: vec![0; 1024 / 4],
        }
    }
}

impl BotRadio {
    pub fn mmio_read(&self, addr: usize) -> u32 {
        // if addr = 0, return the status byte, otherwise fall back to self.memory
    }

    pub fn mmio_write(&mut self, addr: usize, val: u32) {
        // we need support for a `radio_drop(nth: u8)` syscall
        //
        // invoking this drops nth message from `self.memory` and shuffles the
        // following messages to previous places
        //
        // e.g. `radio_drop(0)` removes the first message and then copies the
        // second message into first message's place etc. (can be done quite
        // efficiently with `&[T]::copy_within()`, just remember to zero-out 
        // length of the last message)
    }

    pub fn recv_message(&mut self, msg: &[u32]) {
        // TODO early-reject message if it doesn't match the filter

        let mut ptr = 1;
        
        loop {
            let Some(len) = self.memory.get(ptr) else {
                // No more space in the buffer, welp.
                return;
            };
            
            if len == 0 {
                self.memory[ptr] = len;
                self.memory[ptr + 1..=ptr + len].copy_from_slice(msg);

                // ^ careful, this can panic! 
                // |
                // | if `ptr + len >= self.memory.len()`, there is *some* space
                // | left, but not sufficient to store the entire message
                // |
                // | in this case we could either drop the message or just save
                // | it partially (store as many bytes as there's space left and
                // | call it a day)
                // |
                // | i think the latter is more fun, it makes also the best use
                // | of the buffer
                
                // Message received, yay!
                return;
            } else {
                // There's already a message here, skip over it
                ptr += len + 1;
            }
        }
        
        // Note that since this algorithms walks the entire message list for
        // each received message, it's quite inefficient - can be made O(1) by
        // saving the next free ptr (as in `self.next_free_ptr: usize`) instead
        // of recalculating it all the time
    }
}
```

I think this makes the best usage of the available space, since you can get to 100% utilization. 

> How would we like to deal with message mangling? 

I see the issues - I think it'd be best to postpone this for the time being and go back once the main mechanic is done; maybe a solution will materialize then 😄

Woonters committed:
change message send buffer to splice for quicker code

Woonters committed:
rework message buffer into linear array

Add all the tests as well as intergration with large portions of kartoffel/radio and bot/radio.rs

Woonters committed:
Add internal interactions with radio
